src/toast/ops/totalconvolve.py

```diff
@@ -0,0 +1,695 @@
+    def _check_det_flag_mask(self, proposal):
+        check = proposal["value"]
+        if check < 0:
+            raise traitlets.TraitError("Det flag mask should be a positive integer")
+        return check
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        return
+
+    @property
+    def available(self):
+        """Return True if ducc0.totalconvolve is found in the library search path."""
+        return totalconvolve is not None
+
+    hwp_angle = Unicode(
+        None, allow_none=True, help="Observation shared key for HWP angle"
+    )
+
+    @function_timer
+    def _exec(self, data, detectors=None, **kwargs):
+        if not self.available:
+            raise RuntimeError("ducc0.totalconvolve is not available")
+
+        if self.detector_pointing is None:
+            raise RuntimeError("detector_pointing cannot be None.")
+
+        log = Logger.get()
+
+        timer = Timer()
+        timer.start()
+
+        all_detectors = self._get_all_detectors(data, detectors)
+
+        for det in all_detectors:
+            verbose = self.comm.rank == 0 and self.verbosity > 0
+
+            # Expand detector pointing
+            self.detector_pointing.apply(data, detectors=[det])
+
+            if det in self.sky_file_dict:
+                sky_file = self.sky_file_dict[det]
+            else:
+                sky_file = self.sky_file.format(detector=det, mc=self.mc)
+
+            if det in self.beam_file_dict:
+                beam_file = self.beam_file_dict[det]
+            else:
+                beam_file = self.beam_file.format(detector=det, mc=self.mc)
+
+            lmax, mmax = self.get_lmmax(sky_file, beam_file)
+            sky = self.get_sky(sky_file, lmax, det, verbose)
+            beam = self.get_beam(beam_file, lmax, mmax, det, verbose)
+
+            theta, phi, psi, psi_pol = self.get_pointing(data, det, verbose)
+            pnt = self.get_buffer(theta, phi, psi, det, verbose)
+            del theta, phi, psi
+
+            convolved_data = self.convolve(sky, beam, lmax, mmax, pnt, det, verbose)
+
+            self.calibrate_signal(data, det, beam, convolved_data, verbose)
+            self.save(data, det, convolved_data, verbose)
+
+            del pnt, beam, sky
+
+            if verbose:
+                timer.report_clear(f"totalconvolve process detector {det}")
+
+        return
+
+    def _get_all_detectors(self, data, detectors):
+        """Assemble a list of detectors across all processes and
+        observations in `self._comm`.
+        """
+        my_dets = set()
+        for obs in data.obs:
+            # Get the detectors we are using for this observation
+            obs_dets = obs.select_local_detectors(detectors)
+            for det in obs_dets:
+                my_dets.add(det)
+            # Make sure detector data output exists
+            obs.detdata.ensure(self.det_data, detectors=detectors)
+        all_dets = self.comm.gather(my_dets, root=0)
```

Comment:
Here is another case where we need to allow running without MPI.
```
all_dets = my_dets
if self.comm is not None:
    all_dets = self.comm.gather(my_dets, root=0)
    if self.comm.rank == 0:
        for some_dets in all_dets:
            my_dets.update(some_dets)
        my_dets = sorted(my_dets)
    all_dets = self.comm.bcast(my_dets, root=0)

Reply:
Done.